rotkehlchen/chain/ethereum/adex/adex.py

```diff
@@ -0,0 +1,352 @@
+from functools import partial
+from typing import TYPE_CHECKING, Any, DefaultDict, Dict, List, Optional, Union, overload
+
+from eth_typing.evm import ChecksumAddress
+from eth_utils import to_checksum_address
+from eth_utils.typing import HexAddress, HexStr
+from gevent.lock import Semaphore
+from typing_extensions import Literal
+
+from rotkehlchen.accounting.structures import Balance
+from rotkehlchen.assets.asset import EthereumToken
+from rotkehlchen.chain.ethereum.graph import GRAPH_QUERY_LIMIT, Graph, format_query_indentation
+from rotkehlchen.chain.ethereum.utils import generate_address_via_create2
+from rotkehlchen.errors import RemoteError
+from rotkehlchen.fval import FVal
+from rotkehlchen.inquirer import Inquirer
+from rotkehlchen.premium.premium import Premium
+from rotkehlchen.typing import ChecksumEthAddress, Price, Timestamp
+from rotkehlchen.user_messages import MessagesAggregator
+from rotkehlchen.utils.interfaces import EthereumModule
+
+from .graph import BONDS_QUERY, UNBOND_REQUESTS_QUERY, UNBONDS_QUERY
+from .typing import Bond, PoolBalance, Unbond, UnbondRequest
+from .utils import (
+    ADX_AMOUNT_MANTISSA,
+    CREATE2_SALT,
+    IDENTITY_FACTORY_ADDR,
+    IDENTITY_PROXY_INIT_CODE,
+    POOL_ID_POOL_NAME,
+)
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.ethereum.manager import EthereumManager
+    from rotkehlchen.db.dbhandler import DBHandler
+
+
+class Adex(EthereumModule):
+    """AdEx integration module
+
+    AdEx subgraph:
+    https://github.com/samparsky/adex_subgraph
+    """
+    def __init__(
+            self,
+            ethereum_manager: 'EthereumManager',
+            database: 'DBHandler',
+            premium: Optional[Premium],
+            msg_aggregator: MessagesAggregator,
+    ) -> None:
+        self.ethereum = ethereum_manager
+        self.database = database
+        self.premium = premium
+        self.msg_aggregator = msg_aggregator
+        self.trades_lock = Semaphore()
+        try:
+            self.graph: Optional[Graph] = Graph(
+                'https://api.thegraph.com/subgraphs/name/adexnetwork/adex-protocol',
+            )
+        except RemoteError as e:
+            self.graph = None
+            self.msg_aggregator.add_error(
+                f'Could not initialize the AdEx subgraph due to {str(e)}. '
+                f'All AdEx balances and historical queries are not functioning until this is fixed. '  # noqa: E501
+                f'Probably will get fixed with time. If not report it to Rotkis support channel ',
+            )
+
+    @staticmethod
+    def _calculate_adex_balances(
+            bonds: List[Bond],
+            unbonds: List[Unbond],
+            unbond_requests: List[UnbondRequest],
+            adx_usd_price: Price,
+    ) -> Dict[ChecksumAddress, List[PoolBalance]]:
+        """Given a list of bonds, unbonds and unbond requests return the
+        balance in the protocol (staked amount) per address as <LiquidityPool>).
```

Comment:
LiquidityPool?

Also question. So bond is like a deposit, unbond is withdrawal and unbond request is withdrawal request, right? You need to wait 30 days to unstake and withdraw from what I have seen.